feat: Add error boundaries to layout components and introduce RouteEr…

ngsanthosh opened the pull request:
### Related Issues

- Fixes wso2/product-is#25689

### Purpose

This PR implements route-level error boundaries to isolate errors to specific routes/features in both the Console and MyAccount applications. Previously, when an error occurred in any route, the entire application would become unresponsive due to only having app-level error boundaries. Users had to reload the entire page to recover.

**What this fixes:**
- Errors are now isolated to individual routes instead of crashing the entire application
- Users can navigate to other routes/sections even when one route encounters an error
- Provides clear error messages with retry functionality
- Automatically resets error state when navigating away from errored routes

**Before:** Error in one route → Entire app unresponsive → Requires full page reload

**After:** Error in one route → Only that route shows error → Other routes remain accessible

### Related PRs

- N/A

### Checklist

- [ ] e2e cypress tests locally verified. (for internal contributers)
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Relevant backend changes deployed and verified
- [ ] [Unit tests](/docs/testing/UNIT_TESTING.md) provided. (Add links if there are any)
- [ ] [Integration tests](https://github.com/wso2/product-is/tree/master/modules/integration) provided. (Add links if there are any)

### Security checks

- [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines
- [ ] Ran FindSecurityBugs plugin and verified report
- [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets

## Developer Checklist (Mandatory)

- [ ] Complete the **Developer Checklist** in the related `product-is` issue to track any behavioral change or migration impact.

---

### Technical Details

**New Components:**
- `features/admin.core.v1/components/route-error-boundary.tsx` - Error boundary for console routes
- `apps/myaccount/src/components/shared/route-error-boundary.tsx` - Error boundary for myaccount routes

**Modified Files:**
- Protected route wrappers for both console and myaccount
- All layout files (app, dashboard, default, full-screen) to wrap routes with error boundaries

**Key Features:**
- ✅ Errors isolated to specific routes
- ✅ Navigation remains functional when errors occur
- ✅ Clear error messages using existing i18n translations
- ✅ Automatic error state reset on route change
- ✅ Retry functionality
- ✅ Error logging for debugging
- ✅ No breaking changes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Centralized, consistent error handling across layouts and routes — improved recovery when route/chunk loading fails and preservation of return path for smoother navigation.

* **New Features**
  * Reusable, localized "broken page" fallback screens with illustration and retry/refresh action.
  * Route-level error boundaries applied consistently to protected and public routes with shared, reusable error-handling utilities and reporting.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ngsanthosh commented:
@DonOmalVindula 

Thank you for the detailed review and suggestions!

You're right, I'll address the ESLint and type-check failures.

Regarding the component implementation, that's an excellent point. I'll refactor both RouteErrorBoundary components to be functional components that wrap the existing ErrorBoundary from @wso2is/react-components as you suggested. This will improve code reuse and consistency.

ngsanthosh commented:
@JayaShakthi97 I have cleared the lint issues. Kindly check and do the needful :) 

Thanks.

ngsanthosh commented:
@DonOmalVindula I have made necessary changes, Kindly check.

@brionmario Could you please ask someone from team to check this PR and do the needful.

Thanks in advance.

DonOmalVindula commented:
> @DonOmalVindula Any update? its been very long

Upon testing your change locally, I noticed that the sidepanel collapse/expand functionality

ngsanthosh committed:
refactor: extract duplicated error boundary logic to shared utility module

- Create error-boundary-utils.tsx with createRouteErrorHandler() and createBrokenPageFallback() functions
- Extract sessionStorage error handler logic shared across layouts
- Extract broken page fallback UI component logic shared across layouts
- Update default-layout.tsx, app-layout.tsx, and full-screen-layout.tsx to use shared utilities
- Remove duplicated error handling code and unused imports from all three layouts
- Improves code reuse and maintainability across console layouts

ngsanthosh commented:
wso2 team any update on this?

DonOmalVindula commented on apps/console/src/layouts/dashboard-layout.tsx:
```diff
@@ -194,6 +194,29 @@ const DashboardLayout: FunctionComponent<RouteComponentProps> = (
 
+    const handleRouteChunkError = (_error: Error, _errorInfo: React.ErrorInfo): void => {
+        sessionStorage.setItem("auth_callback_url_console", config.deployment.appHomePath);
+    };
```
+1 for this suggestion. Let's use a common function here

DonOmalVindula commented on apps/myaccount/src/layouts/app.tsx:
```diff
@@ -97,11 +101,21 @@ export const AppLayout: FunctionComponent<Record<string, unknown>> = (): ReactEl
+                                                            <route.component { ...renderProps } />
+                                                        </ErrorBoundary>
+                                                    );
+                                                } }
```
For the moment, this approach is fine

DonOmalVindula commented on apps/console/src/layouts/app-layout.tsx:
```diff
@@ -101,11 +85,22 @@ const AppLayout: FunctionComponent<Record<string, unknown>> = (): ReactElement =
+                                                            <route.component { ...renderProps } />
+                                                        </ErrorBoundary>
+                                                    );
+                                                } }
```
For the moment, this approach is fine

DonOmalVindula commented:
> https://drive.google.com/file/d/1ap2l_Dp84bHTK7_Kr0gkv1ZWPWG6puzO/view?usp=sharing
> 
> Above is a screen recording of the changes done. Kindly check and verify @DonOmalVindula

Hi @ngsanthosh,

Sorry about the delay. The previous bug is now resolved. But now I'm seeing a different bug where the state of the sidepanel resets when i click on a navigation item. The buggy behaviour is shown below


https://github.com/user-attachments/assets/8b18b61a-d6de-412f-b459-4ee2ec8f01a5

Where as in the correct behaviour, the sidepanel should persist it's state when clicking on navigation items as shown below. Please fix it.

https://github.com/user-attachments/assets/4e945e2c-ac96-4846-8c81-4408fa495cdd


Additionally, you have to fix the following.
- Comments generated by the co-pilot review
- ESLint issues shown in the CI/CD Pipeline

ngsanthosh committed:
refactor: address PR review comments

- Add type annotations for handleRouteChunkError const values
- Use shared utilities createRouteErrorHandler and createBrokenPageFallback in dashboard-layout
- Fix copyright years to use range format (2019-2025, 2020-2025)
- Move ROUTE_ERROR_EVENT_ID constant before JSDoc comment
- Remove key prop from ErrorBoundary to fix sidepanel state reset bug
- Fix import order in error-boundary-utils.tsx
- Remove unused imports in default-layout.tsx

ngsanthosh committed:
AI code review changes

Added the key prop using the current pathname to ErrorBoundary and RouteErrorBoundary components across multiple layout and route files. This ensures proper remounting of error boundaries when the route changes, improving error handling and component lifecycle management.
